src/super_gradients/training/utils/quantization/fine_grain_quantization_utils.py

```diff
@@ -0,0 +1,192 @@
+                    quantized_type=child_module.quantized_type,
+                    input_quant_descriptor=child_module.input_quant_descriptor,
+                    weights_quant_descriptor=child_module.weights_quant_descriptor,
+                    action=QuantizedMetadata.ReplacementAction.REPLACE
+                )
+
+            if isinstance(child_module, nn.Module):  # recursive call
+                mapping_instructions.update(self._preprocess_skips_and_custom_mappings(child_module, nesting + (name,)))
+
+        return mapping_instructions
+
+    def _instantiate_quantized_from_float(self, float_module, metadata):
+        base_classes = (QuantMixin, QuantInputMixin, SGQuantMixin)
+        if not issubclass(metadata.quantized_type, base_classes):
+            raise AssertionError(f"Quantization suite for {type(float_module).__name__} is invalid. "
+                                 f"{metadata.quantized_type.__name__} must inherit one of "
+                                 f"{', '.join(map(lambda _: _.__name__, base_classes))}")
+
+        # USE PROVIDED QUANT DESCRIPTORS, OR DEFAULT IF NONE PROVIDED
+        quant_descriptors = dict()
+        if issubclass(metadata.quantized_type, (QuantMixin, QuantInputMixin)):
+            quant_descriptors = {
+                'quant_desc_input': metadata.input_quant_descriptor or self._get_default_quant_descriptor()
+            }
+        if issubclass(metadata.quantized_type, QuantMixin):
+            quant_descriptors.update({
+                'quant_desc_weight': (metadata.weights_quant_descriptor or self._get_default_quant_descriptor())
+            })
+
+        if not hasattr(metadata.quantized_type, 'from_float'):
+            assert isinstance(metadata.quantized_type, SGQuantMixin), \
+                f'{metadata.quantized_type.__name__} must inherit from ' \
+                f'{SGQuantMixin.__name__}, so that it would include `from_float` class method'
+
+        return metadata.quantized_type.from_float(float_module, **quant_descriptors)
+
+    def _maybe_quantize_one_layer(self, module, child_name, nesting, child_module, mapping_instructions) -> bool:
+        # if we don't have any instruction for the specific layer or the specific type - we continue
+        # NOTE! IT IS IMPORTANT TO FIRST PROCESS THE NAME AND ONLY THEN THE TYPE
+        if _imported_pytorch_quantization_failure is not None:
+            raise _imported_pytorch_quantization_failure
+        for candidate_key in ('.'.join(nesting + (child_name,)), type(child_module)):
+            if candidate_key not in mapping_instructions:
+                continue
+            metadata: QuantizedMetadata = mapping_instructions[candidate_key]
+            if metadata.action == QuantizedMetadata.ReplacementAction.SKIP:
+                return True
+            elif metadata.action == QuantizedMetadata.ReplacementAction.UNWRAP:
+                assert isinstance(child_module, SkipQuantization)
```

Comment:
Why the assert ?

Reply:
Just a sanity that we don't unwrap non-wrapper modules